docker,ci,make: add Ceph-Dashboard support and other improvements

epuertat opened the pull request:
This PR is to make it easier for developers to test Ceph-Dashboard when doing integration testing.
Fixes: #357
Fixes: #382

## docker: add Ceph-Dashboard support

Include `ceph-mgr-dashboard` package and deps (grpcio).

Create new ceph-devel service to mount local ceph repo inside the container (for dev purposes).

## docker: add ceph-devel container  
Also:
* Extend healthcheck retries to prevent random failures
* Configure nvmeof gateway in Ceph
* Fix nvmeof-cli aliases (the leading space prevented the shell from
  adding these to the history)
 * Add syntax highlight tag to the Dockerfile.spdk

## docker: add custom Ceph repo
    
* A custom Ceph YUM repo (usually a ceph-ci generated/Chacra one, like
  https://3.chacra.ceph.com/r/ceph/ceph-nvmeof-mon/c1ced49d133c2bcd1fa9311a35688bf641b93c76/centos/9/flavors/default/)
* New make targer `clean_cache` to prune builder cache (required when
  changing the Ceph YUM Repo.
* Repo/RPM GPG check disabled. Not ideal, but ceph-ci builds don't sign
  RPMs.
* Specify ulimits (nofile 1024) for the Ceph container. In Docker CE,
  ceph-mon's ms_dispatch thread iterated through all available file
  descriptors and tried to close them. By setting nofile to 1024, this
  is heavily constrained. This behaviour is not exhibited with
  Fedora's Moby-engine.
 
## docker: support target-arch in SPDK

* Remove exising handling for AVX512*, VPCLMULQDQ and RDSEED
  instructions.
* Replace it with spdk/configure --target-arc=<arch>
 * Tune SPDK's yum/dnf cache and settings.
    
## ci: remove nvmeof-devel dependency

nvmeof-devel is not required for pytests, since they're mounted via `run` CLI command.
    
## make: fix alias issue

Alias was not properly working for `cephnvmf*` as it was erased from the shell history.

## docker: save SPDK target arch as label

SPDK/DPDK relies on pretty modern CPU instruction sets, which renders that incompatible with certain servers (Ivy Bridge and older). This new `SPDK_TARGET_ARCH` setting allows setting that. The new default (`x86-64-v2`) could result in a less performant SPDK.

## ci: update upload-download artifact actions
    
The v4 of these actions provide 90% speed-ups by compressing artifacts.
Migration to this new action means zero cost.

 Additional core clean-up performed.
    
## ci: split ceph job and update checkout

There's no need for `ceph` to be built together with the rest of the nvmeof services. In fact this container could/should be pulled off to a separate repo/CI.    

## ci: cancel CI run when another starts
    
When a user pushes a change to a PR, if there was an ongoing CI run,
this will cancel any previous runs. This will save some CO2.

epuertat committed:
docker: add Ceph-Dashboard support

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat commented on Dockerfile.ceph:
```diff
@@ -1,4 +1,5 @@
 # syntax = docker/dockerfile:1.4
+# vim: syntax=dockerfile
```
BTW this is a trick for Github's linguist tool to properly syntax-highlight files with non-standard extensions (e.g.: `.ceph` for a `Dockerfile`).

epuertat commented on Dockerfile.ceph:
```diff
@@ -88,6 +102,12 @@ USER ceph
 WORKDIR /ceph
 ADD --chown=ceph:ceph --chmod=755 https://raw.githubusercontent.com/ceph/ceph/v$CEPH_CLUSTER_VERSION/src/vstart.sh .
 
+COPY <<EOF ./CMakeCache.txt
+ceph_SOURCE_DIR:STATIC=/ceph
+WITH_MGR_DASHBOARD_FRONTEND:BOOL=ON
+WITH_RBD:BOOL=ON
+EOF
```
vstart assumes that Ceph is always run from sources, and hence the CMake artifacts are over there. We should probably override that if some `WITH_MGR_DASHBOARD_FRONTEND` env var is defined.

baum reviewed (approved):
lgtm 🖖

epuertat committed:
docker: add ceph-devel container

Also:
* Extend healthcheck retries to prevent random failures
* Configure nvmeof gateway in Ceph
* Fix nvmeof-cli aliases (the leading space prevented the shell from
  adding these to the history)
* Add syntax highlight tag to the Dockerfile.spdk

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat commented on mk/misc.mk:
```diff
@@ -5,6 +5,7 @@ NVMEOF_CLI = $(DOCKER_COMPOSE_ENV) $(DOCKER_COMPOSE) run --rm nvmeof-cli --serve
 
 alias: ## Print bash alias command for the nvmeof-cli. Usage: "eval $(make alias)"
-	@echo alias cephnvmf=\"$(NVMEOF_CLI)\" \; alias cephnvmf-ipv6=\'$(NVMEOF_CLI_IPV6)\'
+	@echo alias cephnvmf=\"$(strip $(NVMEOF_CLI))\"
+	@echo alias cephnvmf-ipv6=\"$(strip $(NVMEOF_CLI_IPV6))\"
```
`make` doesn't really care about leading whitespaces, and expressions like:
```make
NVMEOF_CLI := $(DOCKER_COMPOSE_ENV) ...
```
result in:
```sh
> make alias
alias cephnvmf=" docker-compose ..."
```
This whitespace before `docker-compose` matches a convention in most shell interpreters, where commands with a leading whitespace are explicitly excluded from the shell history (e.g.: running a command with a a secret/password).

epuertat committed:
docker: add custom Ceph repo

* A custom Ceph YUM repo (usually a ceph-ci generated/Chacra one, like
  https://3.chacra.ceph.com/r/ceph/ceph-nvmeof-mon/c1ced49d133c2bcd1fa9311a35688bf641b93c76/centos/9/flavors/default/)
* New make targer `clean_cache` to prune builder cache (required when
  changing the Ceph YUM Repo.
* Repo/RPM GPG check disabled. Not ideal, but ceph-ci builds don't sign
  RPMs.
* Specify ulimits (nofile 1024) for the Ceph container. In Docker CE,
  ceph-mon's ms_dispatch thread iterated through all available file
  descriptors and tried to close them. By setting nofile to 1024, this
  is heavily constrained. This behaviour is not exhibited with
  Fedora's Moby-engine.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat commented on .env:
```diff
@@ -57,7 +57,9 @@ SPDK_CENTOS_REPO_VER="9.0-21.el9"
 # Ceph Cluster
 CEPH_CLUSTER_VERSION="${CEPH_VERSION}"
-CEPH_VSTART_ARGS="--without-dashboard --memstore"
+# CEPH_CLUSTER_CEPH_REPO_BASEURL= # Uncomment for specifying a custom Ceph YUM repo
```
@baum uncomment and set this to the ceph-ci repo URL:
```suggestion
CEPH_CLUSTER_CEPH_REPO_BASEURL="https://3.chacra.ceph.com/r/ceph/ceph-nvmeof-mon/c1ced49d133c2bcd1fa9311a35688bf641b93c76/centos/9/flavors/default/"
```
You'll need to run `make clean_cache` to prune the build cache.

epuertat committed:
docker: support target-arch in SPDK

* Remove exising handling for AVX512*, VPCLMULQDQ and RDSEED
  instructions.
* Replace it with spdk/configure --target-arc=<arch>
* Tune SPDK's yum/dnf cache and settings.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat commented:
@gbregman please review this PR as it removes part of the fixes you introduced for the missing CPU instruction.

@baum I pulled off the `--target-arch` configure param to a separate env. var, in order to make it easier to generate multiple SPDK versions from matrix job (e.g.: [`x86-64-v2` and `x86-64-v4`](https://developers.redhat.com/blog/2021/01/05/building-red-hat-enterprise-linux-9-for-the-x86-64-v2-microarchitecture-level#background_of_the_x86_64_microarchitecture_levels) might be enough to cover legacy and modern CPUs). Probably not required in every build, but just when publishing a release.

gbregman reviewed (approved):
LGTM, only I don't really know much about these files. One comment though, if you removed the SPDK_DISABLE_VPCLMULQDQ and SPDK_DISABLE_AVX512 vars, shouldn't you also remove them from .github/workflows/build-container.yml?

gbregman commented:
@epuertat notice that on top of disabling the AVX512 and the MULQDQ instructions we also disable RDSEED. Only this id hardcoded in Dockerfile.spdk:
```
# Disable RDSEED, see https://github.com/ceph/ceph-nvmeof/issues/259
RUN \
    sed -i "s/^\( \+'RDSEED'\)/#\1/" dpdk/config/x86/meson.build
```

epuertat committed:
ci: remove nvmeof-devel dependency

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
make: fix alias issue

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
docker: save SPDK target arch as label

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -1,11 +1,15 @@
+---
+# yamllint disable rule:line-length
```
I run this through `yamllinter` and it reported several (cosmetic) issues.

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -1,11 +1,15 @@
   pull_request:
+    branches:
+      - devel
+  schedule:
+    - cron: '0 0 * * *'
```
I'm recovering the nightly runs.

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -22,28 +26,37 @@ jobs:
 
-      - name: Build container images
-        run: make build SPDK_DISABLE_VPCLMULQDQ="yes" SPDK_DISABLE_AVX512="yes"
+      - name: Build container images - ceph
+        run: make build SVC=ceph SPDK_TARGET_ARCH=x86-64-v2
+
+      - name: Build container images - spdk
+        run: make build SVC="spdk" SPDK_TARGET_ARCH=x86-64-v2
+
+      - name: Build container images - bdevperf
+        run: make build SVC="bdevperf" SPDK_TARGET_ARCH=x86-64-v2
+
+      - name: Build container images - nvmeof
+        run: make build SVC="nvmeof" SPDK_TARGET_ARCH=x86-64-v2
+
+      - name: Build container images - nvmeof-cli
+        run: make build SVC="nvmeof-cli" SPDK_TARGET_ARCH=x86-64-v2
+
+      - name: Build container images - discovery
+        run: make build SVC="discovery" SPDK_TARGET_ARCH=x86-64-v2
```
I have to check if setting `SPDK_TARGET_ARCH` as env. variable can spare us from repeating it here.

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -22,28 +26,37 @@ jobs:
-          docker save $QUAY_CEPH:$CEPH_VERSION  > vstart-cluster.tar
-          docker save bdevperf > bdevperf.tar
-          docker save nvmeof-devel > nvmeof-devel.tar
+          docker save $QUAY_NVMEOF:$NVMEOF_VERSION -o nvmeof.tar.xz
+          docker save $QUAY_NVMEOFCLI:$NVMEOF_VERSION -o nvmeof-cli.tar.xz
+          docker save $QUAY_CEPH:$CEPH_VERSION -o ceph.tar.xz
+          docker save bdevperf -o bdevperf.tar.xz
```
`xz` (slightly better than bzip2) compression should improve things here.

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -448,9 +454,9 @@ jobs:
+        uses: andstor/file-existence-action@20b4d2e596410855db8f9ca21e96fbe18e12930b  # v2, pinned to SHA for security reasons
         with:
-            files: "/tmp/coredump/core.*"
+          files: "/tmp/coredump/core.*"
```
BTW @gbregman I recently used `systemd-coredump` and it's geat. Coredumps inside a container are automatically available to the host's `coredumpctl` command, and you can easily debug as if you were doing it inside:
```
CONTAINER_ID=#<get from docker ps>
SYSROOT=$(docker inspect -f $'{{.GraphDriver.Data.MergedDir}}' $CONTAINER_ID)

# Let's launch GDB
gdb -iex "set sysroot $SYSROOT" $SYSROOT/usr/local/bin/nvmf_tgt /tmp/core
```

epuertat commented on docker-compose.yaml:
```diff
@@ -53,34 +52,47 @@ services:
-      sh -c './vstart.sh --new $$VSTART_ARGS &&
+      sh -c './vstart.sh --new $$CEPH_VSTART_ARGS &&
       ceph osd pool create rbd &&
+      echo ceph dashboard nvmeof-gateway-add -i <(echo nvmeof-devel:5500) nvmeof.1 &&
```
This is currently useless. It should be moved (without the `echo`, of course) into the vstart.sh when invoked with a new flag, like `vstart.sh --nvmf` (note to @nizamial09 ), which would also create the `rbd` pool.

epuertat committed:
ci: update upload-download artifact actions

The v4 of these actions provide 90% speed-ups by compressing artifacts.
Migration to this new action means zero cost.

Additional core clean-up performed.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: split ceph job and update checkout

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

epuertat committed:
ci: cancel CI run when another starts

When a user pushes a change to a PR, if there was an ongoing CI run,
this will cancel any previous runs. This will save some CO2.

Signed-off-by: Ernesto Puerta <epuertat@redhat.com>
